File: 525-contiguous-array/525-contiguous-array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& arr) {
        
        unordered_map<int, int> mp; // sum -> index
        mp[0] = -1;
        
        int len=0;
        int sum=0;
        
        for(int i=0; i<arr.size(); i++){
            
            if(arr[i] == 0){
                sum += -1;
            }
            else if(arr[i] == 1){
                sum += 1;
            }
            
            if(mp.find(sum) != mp.end()){
                int curLen = i - mp[sum];
                len = max(curLen, len);
            }
            else{
                mp[sum] = i;
            }
        }
        
        return len;
    }
};

File: 525-contiguous-array/525-contiguous-array.java
class Solution {
    public int findMaxLength(int[] nums) {
        
        Map<Integer, Integer> map = new HashMap<>();
        int maxlen = 0;
        int sum=0;
        
        map.put(0, -1);
        for(int i=0; i<nums.length; i++){
            if(nums[i]==0) sum += -1;
            else sum += 1;
            
            if(map.containsKey(sum)==true){
                int len = i - map.get(sum);
                maxlen = Math.max(len, maxlen);
            }
            else{
                map.put(sum, i);
            }
        }
        return maxlen;
    }
}